### PR TITLE
feat: bandit evaluation

### PR DESCRIPTION
**Pull request by greghuels**
## Summary

  Implements contextual bandit evaluation for the Eppo C++ SDK, enabling dynamic action selection based on subject and action attributes using linear models with exploration.
  
  Note: Bandit logging deduplication is implemented in [this PR](https://github.com/Eppo-exp/cpp-sdk/pull/6).

  ## Changes

  ### Core Bandit Model (`src/bandit_model.hpp/cpp`)
  - **Data structures** for bandit configuration:
    - `BanditNumericAttributeCoefficient` - Linear coefficients for numeric features
    - `BanditCategoricalAttributeCoefficient` - One-hot encoded coefficients for categorical features
    - `BanditCoefficients` - Action-specific coefficients (intercept + subject/action attributes)
    - `BanditModelData` - Complete model with gamma (exploration), default scores, and action probability floor
    - `BanditConfiguration` - Model metadata and configuration
    - `BanditVariation` - Links bandits to feature flag variations
  - Full JSON serialization/deserialization support with ISO 8601 timestamp handling

  ### Bandit Evaluation Engine (`src/evalbandits.hpp/cpp`)
  - **Context attribute handling**:
    - `ContextAttributes` - Separates numeric and categorical attributes
    - Type inference from generic `Attributes`
  - **Scoring algorithm**:
    - Linear model scoring for actions using subject and action attributes
    - Handles missing attributes via `missingValueCoefficient`
  - **Action selection**:
    - Softmax-based probability calculation with gamma parameter for exploration
    - Action probability floor to ensure minimum exploration
    - Deterministic shard-based selection (10,000 shards)
    - Optimality gap calculation for regret analysis
  - **Telemetry**: `BanditEvent` structure with action probabilities, attributes, and metadata

  ### Client Integration (`src/client.hpp/cpp`)
  - `BanditLogger` interface for action logging
  - `getBanditAction()` method:
    1. Evaluates feature flag assignment
    2. Retrieves associated bandit configuration
    3. Scores and selects action using contextual attributes
    4. Logs action with probability, optimality gap, and model version
  - Returns `BanditResult` with both variation and selected action

  ### Configuration Management
  - `Configuration` class extended to store bandit configurations
  - O(1) lookup via `banditFlagAssociations_` map (flag key → variation value → bandit)
  - `getBanditVariant()` and `getBanditConfiguration()` accessors

  ### Testing
  - **496 lines** in `test_bandit_evaluation.cpp` - scoring, selection, edge cases
  - **271 lines** in `test_bandit_model.cpp` - JSON serialization round-trip tests
  - **159 lines** in `test_configuration.cpp` - configuration storage and lookup
  - **UFC test data**: 8 bandit test scenarios including cold start, missing actions, and UK targeting
  - Test data files: `bandit-flags-v1.json`, `bandit-models-v1.json`

  ## Technical Details

  **Algorithm**: Contextual bandit with softmax action selection
  - Score = intercept + Σ(coefficient × attribute_value)
  - Probability ∝ exp(score / gamma) with minimum floor
  - Deterministic assignment using subject key sharding

  **Performance**: O(1) configuration lookup, O(n) action scoring where n = number of actions

## Example Usage  
```cpp
#include <iostream>
#include <fstream>
#include <memory>
#include <nlohmann/json.hpp>
#include "client.hpp"

// Simple console-based assignment logger
class ConsoleAssignmentLogger : public eppoclient::AssignmentLogger {
public:
    void logAssignment(const eppoclient::AssignmentEvent& event) override {
        std::cout << "\n=== Assignment Log ===" << std::endl;
        std::cout << "Experiment: " << event.experiment << std::endl;
        std::cout << "Feature Flag: " << event.featureFlag << std::endl;
        std::cout << "Allocation: " << event.allocation << std::endl;
        std::cout << "Variation: " << event.variation << std::endl;
        std::cout << "Subject: " << event.subject << std::endl;
        std::cout << "Timestamp: " << event.timestamp << std::endl;

        if (!event.subjectAttributes.empty()) {
            std::cout << "Subject Attributes:" << std::endl;
            for (const auto& [key, value] : event.subjectAttributes) {
                std::cout << "  " << key << ": ";
                std::visit([](const auto& v) {
                    using T = std::decay_t<decltype(v)>;
                    if constexpr (std::is_same_v<T, std::monostate>) {
                        std::cout << "(none)";
                    } else if constexpr (std::is_same_v<T, bool>) {
                        std::cout << (v ? "true" : "false");
                    } else {
                        std::cout << v;
                    }
                }, value);
                std::cout << std::endl;
            }
        }

        if (!event.metaData.empty()) {
            std::cout << "Metadata:" << std::endl;
            for (const auto& [key, value] : event.metaData) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }
        std::cout << "=====================\n" << std::endl;
    }
};

// Simple console-based bandit logger
class ConsoleBanditLogger : public eppoclient::BanditLogger {
public:
    void logBanditAction(const eppoclient::BanditEvent& event) override {
        std::cout << "\n=== Bandit Action Log ===" << std::endl;
        std::cout << "Flag Key: " << event.flagKey << std::endl;
        std::cout << "Bandit Key: " << event.banditKey << std::endl;
        std::cout << "Subject: " << event.subject << std::endl;
        std::cout << "Action: " << event.action << std::endl;
        std::cout << "Action Probability: " << event.actionProbability << std::endl;
        std::cout << "Optimality Gap: " << event.optimalityGap << std::endl;
        std::cout << "Model Version: " << event.modelVersion << std::endl;
        std::cout << "Timestamp: " << event.timestamp << std::endl;

        if (!event.subjectNumericAttributes.empty()) {
            std::cout << "Subject Numeric Attributes:" << std::endl;
            for (const auto& [key, value] : event.subjectNumericAttributes) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }

        if (!event.subjectCategoricalAttributes.empty()) {
            std::cout << "Subject Categorical Attributes:" << std::endl;
            for (const auto& [key, value] : event.subjectCategoricalAttributes) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }

        if (!event.actionNumericAttributes.empty()) {
            std::cout << "Action Numeric Attributes:" << std::endl;
            for (const auto& [key, value] : event.actionNumericAttributes) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }

        if (!event.actionCategoricalAttributes.empty()) {
            std::cout << "Action Categorical Attributes:" << std::endl;
            for (const auto& [key, value] : event.actionCategoricalAttributes) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }

        if (!event.metaData.empty()) {
            std::cout << "Metadata:" << std::endl;
            for (const auto& [key, value] : event.metaData) {
                std::cout << "  " << key << ": " << value << std::endl;
            }
        }
        std::cout << "========================\n" << std::endl;
    }
};

// Simple console-based application logger
class ConsoleApplicationLogger : public eppoclient::ApplicationLogger {
public:
    void debug(const std::string& message) override {
        std::cout << "[DEBUG] " << message << std::endl;
    }

    void info(const std::string& message) override {
        std::cout << "[INFO] " << message << std::endl;
    }

    void warn(const std::string& message) override {
        std::cout << "[WARN] " << message << std::endl;
    }

    void error(const std::string& message) override {
        std::cerr << "[ERROR] " << message << std::endl;
    }
};

// Helper function to load flags configuration from JSON file
eppoclient::ConfigResponse loadFlagsConfiguration(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open flags configuration file: " + filepath);
    }

    nlohmann::json j;
    file >> j;

    eppoclient::ConfigResponse response = j;
    return response;
}

// Helper function to load bandit models from JSON file
eppoclient::BanditResponse loadBanditModels(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open bandit models file: " + filepath);
    }

    nlohmann::json j;
    file >> j;

    eppoclient::BanditResponse response = j;
    return response;
}

int main() {
    try {
        // Load the bandit flags and models configuration
        std::cout << "Loading bandit flags and models configuration..." << std::endl;
        eppoclient::ConfigResponse banditFlags = loadFlagsConfiguration("bandit-flags-v1.json");
        eppoclient::BanditResponse banditModels = loadBanditModels("bandit-models-v1.json");

        // Create configuration store and set the configuration with both flags and bandits
        auto configStore = std::make_shared<eppoclient::ConfigurationStore>();
        configStore->setConfiguration(eppoclient::Configuration(banditFlags, banditModels));

        // Create assignment logger, bandit logger, and application logger
        auto assignmentLogger = std::make_shared<ConsoleAssignmentLogger>();
        auto banditLogger = std::make_shared<ConsoleBanditLogger>();
        auto applicationLogger = std::make_shared<ConsoleApplicationLogger>();

        // Create EppoClient with all parameters including bandit logger
        auto client = std::make_shared<eppoclient::EppoClient>(configStore, assignmentLogger, banditLogger, applicationLogger);

        std::cout << "\n=== Car Bandit: Selecting Car to Recommend ===" << std::endl;

        // Define subject attributes (user/context attributes)
        // Car bandit doesn't require subject attributes, but we can include them
        eppoclient::ContextAttributes subjectAttrs;

        // Define available actions (cars) with their attributes
        std::map<std::string, eppoclient::ContextAttributes> actions;

        // Action 1: Toyota
        // Attributes: speed (numeric)
        eppoclient::ContextAttributes toyotaAction;
        toyotaAction.numericAttributes["speed"] = 120.0;
        actions["toyota"] = toyotaAction;

        // Action 2: Honda (no specific model, but adding for variety)
        eppoclient::ContextAttributes hondaAction;
        hondaAction.numericAttributes["speed"] = 115.0;
        actions["honda"] = hondaAction;

        // Get bandit action recommendation
        eppoclient::BanditResult banditResult = client->getBanditAction(
            "car_bandit_flag",     // flag key
            "user-abc123",         // subject key
            subjectAttrs,          // subject attributes
            actions,               // available actions
            "car_bandit"           // default variation if flag not found
        );

        std::cout << "Bandit selected variation: " << banditResult.variation << std::endl;
        if (banditResult.action.has_value()) {
            std::cout << "Recommended car: " << banditResult.action.value() << std::endl;

            // Use the recommended action in your application logic
            if (banditResult.action.value() == "toyota") {
                std::cout << "✓ Recommending Toyota to user" << std::endl;
            } else if (banditResult.action.value() == "honda") {
                std::cout << "✓ Recommending Honda to user" << std::endl;
            } else {
                std::cout << "✓ Recommending " << banditResult.action.value() << " to user" << std::endl;
            }
        } else {
            std::cout << "No action selected (using default)" << std::endl;
        }

        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}
```

## Example Output
Screenshot 2025-11-05 at 11 28 10 AM

### CODE REVIEW

File: test/test_bandit_evaluation.cpp [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+
+    auto banditLogger = std::make_shared<MockBanditLogger>();
+    auto client = std::make_shared<EppoClient>(configStore, nullptr, banditLogger, nullptr);
+
+    // Load all bandit test cases
+    std::string testCasesDir = "test/data/ufc/bandit-tests";
+    std::vector<BanditTestCase> testCases;
+
+    REQUIRE_NOTHROW(testCases = loadAllBanditTestCases(testCasesDir));
+    REQUIRE(testCases.size() > 0);
```
Comment (aarsilv) [inline test/test_bandit_evaluation.cpp]: 👍

**Review by aarsilv — COMMENTED** [review]
Sorry I'm late to the party but nice work getting bandits in!
